services/relayer/internal/sdk/executor/executor_impl_v1.go

```diff
@@ -0,0 +1,415 @@
+package executor
+
+/*
+ * Copyright 2021 ConsenSys Software Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/consensys/gpact/messaging/relayer/internal/contracts/functioncall"
+	"github.com/consensys/gpact/messaging/relayer/internal/logging"
+	"github.com/consensys/gpact/messaging/relayer/internal/sdk/chainap"
+	"github.com/consensys/gpact/messaging/relayer/internal/sdk/msgstore"
+	"github.com/consensys/gpact/messaging/relayer/internal/sdk/treenode"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// timeout for getting a signed event.
+const eventTimeout = 30 * time.Second
+
+// Function sigs.
+var startFuncSig = [32]byte{0x77, 0xda, 0xb6, 0x11, 0xad, 0x9a, 0x24, 0xb7, 0x63, 0xe2, 0x74, 0x2f, 0x57, 0x74, 0x9a, 0x02, 0x27, 0x39, 0x3e, 0x0d, 0xa7, 0x62, 0x12, 0xd7, 0x4f, 0xce, 0xb3, 0x26, 0xb0, 0x66, 0x14, 0x24}
+var segmentFuncSig = [32]byte{0xb0, 0x15, 0x57, 0xf1, 0xf6, 0x34, 0xb7, 0xc5, 0x07, 0x2a, 0xb5, 0xe3, 0x6d, 0x07, 0xa2, 0x35, 0x5e, 0xf8, 0x19, 0xfa, 0xca, 0x5a, 0x3d, 0x32, 0x14, 0x30, 0xd7, 0x19, 0x87, 0x15, 0x5b, 0x8f}
+var rootFuncSig = [32]byte{0xe6, 0x76, 0x3d, 0xd9, 0x9b, 0xf8, 0x94, 0xd7, 0x2f, 0x34, 0x99, 0xdd, 0x57, 0x2a, 0xa4, 0x28, 0x76, 0xea, 0xe7, 0xae, 0x02, 0x8c, 0x32, 0xff, 0xf2, 0x16, 0x54, 0xe1, 0xbb, 0xc4, 0xc8, 0x07}
+
```

Comment:
It would be good to put in the known limitations of the implementation as a comment at the top of the file. It appears that this implementation will only handle root->one or more segments, but then the segments can't call other segments.

Reply:
I can see that segments calling segments is supported.

Reply:
However, the nonce management (fetching nonce from the blockchain) and gas pricing (asking the blockchain for the recommended gas price) should be highlighted.

Reply:
sounds good